zipline/modelling/engine.py

```diff
@@ -0,0 +1,443 @@
+    """
+    Build a dependency graph containing the given terms and their dependencies.
+
+    Parameters
+    ----------
+    terms : iterable
+        An iterable of zipline.modelling.term.Term.
+
+    Returns
+    -------
+    dependencies : networkx.DiGraph
+        A directed graph representing the dependencies of the desired inputs.
+
+        Each node in the graph has an `extra_rows` attribute, indicating how
+        many, if any, extra rows we should compute for the node.  Extra rows
+        are most often needed when a term is an input to a rolling window
+        computation.  For example, if we compute a 30 day moving average of
+        price from day X to day Y, we need to load price data for the range
+        from day (X - 29) to day Y.
+    """
+    dependencies = DiGraph()
+    parents = set()
+    for term in terms:
+        _add_to_graph(
+            term,
+            dependencies,
+            parents,
+            extra_rows=0,
+        )
+        # No parents should be left between top-level terms.
+        assert not parents
+    return dependencies
+
+
+def _add_to_graph(term,
+                  dependencies,
+                  parents,
+                  extra_rows):
+    """
+    Add the term and all its inputs to dependencies.
+    """
+    # If we've seen this node already as a parent of the current traversal,
+    # it means we have an unsatisifiable dependency.  This should only be
+    # possible if the term's inputs are mutated after construction.
+    if term in parents:
+        raise CyclicDependency(term)
+    parents.add(term)
+
+    try:
+        existing = dependencies.node[term]
+    except KeyError:
+        # We're not yet in the graph: add the term with the specified number of
+        # extra rows.
+        dependencies.add_node(term, extra_rows=extra_rows)
+    else:
+        # We're already in the graph because we've been traversed by
+        # another parent.  Ensure that we have enough extra rows to satisfy
+        # all of our parents.
+        existing['extra_rows'] = max(extra_rows, existing['extra_rows'])
+
+    for subterm in term.inputs:
+        _add_to_graph(
+            subterm,
+            dependencies,
+            parents,
+            extra_rows=extra_rows + term.extra_input_rows,
+        )
+        dependencies.add_edge(subterm, term)
+
+    parents.remove(term)
+
+
+class FFCEngine(with_metaclass(ABCMeta)):
+
+    @abstractmethod
+    def factor_matrix(self, terms, start_date, end_date):
+        """
+        Compute the full factor matrix with a first date of `start`.
+
+        Parameters
+        ----------
+        terms : dict
+            Map from str -> zipline.modelling.term.Term.
+        start : datetime
+            The first date of the matrix.
+        end : datetime
+            The last date of the matrix.
+
+        Returns
+        -------
+        matrix : pd.DataFrame
+            A matrix of factors
+        """
+        raise NotImplementedError()
```

Comment:
I prefer to put the name of the not implemented function here to aid debugging. This is style.